air_quality/air_quality.ino

```diff
@@ -52,106 +78,207 @@ airQualityData_t airData = { 0.0 };
+ *                ratio is used with a quadratic fit function to get an estimate of
+ *                dust particle count (provided by Many Labs). This dust particle
+ *                count is finally combined with humidity data and an accepted
+ *                formula to get PM2.5 values.
+ * =====================================================================================
+ */
+void loop(void)
+{
+  unsigned long time = millis();
+  if (time - lastAvgTime > AVERAGING_MS) {
 
-  }else if(humidityInt >= 60 && humidityInt <= 69){ //Humidity between 60-69%
+    // Compute low pulse occupancy - multiply by 1000 to convert to microseconds
+    airData.dustDetectedRatio = totalPulseDuration / (float)((time - lastAvgTime) * 1000);
+    totalPulseDuration = 0;
+    lastAvgTime = time;
 
-      conc25 = conc25First*(humidity/100)*factor60;
+    // This count estimate comes from co-locating the PPD60 with a Dylos DC1700
+    // and fitting a quadratic curve to the data
+    airData.count = 79740.9269 * (airData.dustDetectedRatio * airData.dustDetectedRatio) +
+                    10175.6562 * airData.dustDetectedRatio + 213.9981;
 
-  }else if(humidityInt >= 70 && humidityInt <= 79){ //Humidity between 70-79%
+    float humidity = 50.0; // Replace with a real humidity sensor value
+    airData.pm25 = dustCountToPM25(airData.count, humidity);
 
-      conc25 = conc25First*(humidity/100)*factor70;
+    // Save values to SD card
+    logString(valuesToCSV(NULL, airData.dustDetectedRatio, 3, lastAvgTime));
 
-  }else if(humidityInt >= 80 && humidityInt <= 89){ //Humidity between 80-89%
+    // Print values to Serial console for debugging
+    serialConnection.print(airData.dustDetectedRatio, 4);
+    serialConnection.print("\t\t");
+    serialConnection.print(airData.count, 4);
+    serialConnection.print("\t");
+    serialConnection.println(airData.pm25, 4);
+  }
+}
 
-      conc25 = conc25First*(humidity/100)*factor80;
 
-  }else if(humidityInt >= 90 && humidityInt <= 100){ //Humidity between 90-100%
+/*
+ * ===  FUNCTION  ======================================================================
+ *         Name:  timeDustPulse
+ *  Description:  Times the duration of a dust pulse and adds the value to accumulator
+ *                variables. Called everytime the interrupt pin changes (low -> high
+ *                or high -> low)
+ * =====================================================================================
+ */
+void timeDustPulse(void)
+{
+  // Check if Pin 2 is high
+  if (digitalRead(2) == HIGH) {
+    // Compute duration
+    // NOTE: micros() overflows every ~70 minutes on Arduino. That could introduce
+    // some error here, although with averaging it's probably a non issue
+    totalPulseDuration += micros() - pulseStart;
+  } else {
+    // Start timing when it's low
+    pulseStart = micros();
+  }
+}
+
 
-      conc25 = conc25First*(humidity/100)*factor90;
+/*
```

Comment:
I'm a little surprised this works/compiles. In a normal C source file, you can't use locally defined function until it has been declared, so you either have to do function declarations before the definitions (e.g. `float dustCountToPM25(float dustCount, float humidity);` at the top of the file), or use header files. The Arduino IDE must run the ino file through a preprocessor that creates declarations for all functions before the definitions, because this seems to compile fine in the IDE...

Still probably not a best practice though, given it wouldn't work in normal C.